Dev: flash additional partitions

tsterbak opened the pull request:
This PR is based on #220 (by @anon1892) and aims to integrate the code in the Installer.

- Include the commands `fastboot_flash_recovery` and `fastboot_reboot_recovery` (thanks to @anon1892)
- Add support for Mi439 (Redmi 7A & co.) (thanks to @anon1892)
- Display device specific notes at the image selection step (thanks to @anon1892)

Other fixes & additions:
- Display the scroll bar by default if scrolling is necessary
- An info text box for additional images

Todo:
- [x] When the "custom recovery already flashed" button is selected, user should not have to select recovery file, as in this PR for now.
- [x] User should be able not to select additional partitions, even on Android 13 (maybe add a button like "custom recovery already flashed".
- [x] When recovery flashing is disabled, `adb reboot recovery` is not performed. Fix that.

tsterbak commented:
@anon1892 I changed a few things and improved the layout in the app a bit.

Most notably, the process now checks for the android version of the ROM you want to flash and only then decides if the selection for additional images should be displayed.

What is missing now:
- [ ] A bit more explanation of why the partitions are needed and where to get the images.
- [ ] Potentially an option in the config to select from which Android version on the additional images should be flashed. This might be necessary for each partition independently (What would you say?).
- [ ] Some in depth testing. I sadly don't have too many devices available right now that need this feature.

Please have a look and give it a test if you can/want :blush: Also, feel free to add a PR to this PR to improve it!

anon1892 commented:
Ok, will test it when I have time.
>  A bit more explanation of why the partitions are needed and where to get the images.

Maybe a button opening an info box?

> Potentially an option in the config to select from which Android version on the additional images should be flashed. 

As far I know, it's only needed for A13. So, no need for an additional config option?
But may we add a checkbox to allow flashing it if the user knows what he is doing ? Just in case...

tsterbak commented:
> Ok, will test it when I have time.

Great! Thanks a lot!
Maybe we could already put some Xiaomi device without OFox requirement in for testing? I will add a Mi 9 Lite, which I have available for testing. But it only requires to wipe with super_emtpy. But its a start :)

> Maybe a button opening an info box?

Yeah, I guess so. I will try to write a good text for it.

> As far I know, it's only needed for A13. So, no need for an additional config option?
But may we add a checkbox to allow flashing it if the user knows what he is doing ? Just in case...

That's a good idea! I will think about a UI layout for that :)

anon1892 commented:
> Maybe we could already put some Xiaomi device without OFox requirement in for testing?

Without OFox for now? It will require more testing, because when OFox support will be added, we will need to test OFox AND TWRP just in case it breaks something, isn't it?

For `dtbo`, `vbmeta` and `super_empty` flashing, there is `Mi439` #226.
And for testing if everything is good without additional partitions, maybe `garden` #224?

anon1892 commented:
Why "recovery already flashed" and "bootloader already unlocked" toggles don't behave the same way?
The first can be clicked before choosing device (so, in reading order), but the second can't be clicked before. Why?

And I can't test properly, I have to make changes in device config, add #218, and maybe others things too...
Why don't you commit all these PRs into `dev` branch? As it is development branch, it can be broken and thr average user should not use it, isn't it?

tsterbak commented:
> Why "recovery already flashed" and "bootloader already unlocked" toggles don't behave the same way?

good point, actually, I guess they both should be only available after the device is detected. What do you think?

> And I can't test properly, I have to make changes in device config, add https://github.com/openandroidinstaller-dev/openandroidinstaller/pull/218, and maybe others things too...
Why don't you commit all these PRs into dev branch? As it is development branch, it can be broken and thr average user should not use it, isn't it?

I understand and I'm sorry for the inconvenience.

I would prefer not to push out releases with too many changes at the same time and also the OFox integration needs some more thought to make it future proof. Being able to flash additional partitions would be valuable immediately and would enable supporting more devices.
I know this increases the efforts for testing but it also reduces the things that can go wrong with each release quite a bit. And in the Android ecosystem, unexpected side effects of changes are common and overlooking things is easy. So I'd rather take this higher effort, lower stress route. 

So, I guess, we would add devices with TWRP support first to test flashing of additional partitions and then move on to add Ofox support.

anon1892 commented:
> good point, actually, I guess they both should be only available after the device is detected. What do you think?

I would say the opposite, or move these buttons to the bottom (I read from top to bottom, so I click on buttons from top to bottom too, and I think a lot of users do the same)

> So I'd rather take this higher effort, lower stress route.

Understandable! But we will need to add `fastboot_flash_recovery` even without adding OFox, since it's a TWRP command and needed for Xiaomi (and maybe others brands) phones to be supported. And I should be able to test this PR with #226

tsterbak commented:
Hey @anon1892, I guess merged and integrated all the required parts for a first test. Let me know if it works for you and/or if something is missing! :)
It will be great to have this merged!

anon1892 commented:
Nice! I will test these changes in a few days.

anon1892 commented:
Hello! I started to test today. I didn't succeed flashing the phone for now.

Some points to enhance : 
- For Xiaomi phones (`brand: xiaomi`), keep https://xiaomifirmwareupdater.com/ instead of adding miui/DEVICE_CODENAME, because we support many devices per config files, and for example "mi439" leads to 404 error (because it's a group name, not a phone code name)
- When the "custom recovery already flashed" button is selected, user should not have to select recovery file, as in this PR for now.
- `mi439` (not capitalized) should be added in supported code list too
- recovery_works_with_device function should validate all codenames of the device config file (in this case, mi439 or pine too)
- User should be able not to select additional partitions, even on Android 13 (maybe add a button like "custom recovery already flashed"

Bug : 
I select good `vbmeta.img` file, but OAI never recognize it. Logs : 
```
INFO     | views.select_view:pick_vbmeta_result:560 - Selected vbmeta from /home/******/vbmeta.img
INFO     | utils:image_works_with_device:51 - Image works with device: ['pine', 'olive', 'olivelite', 'olivewood', 'olives', 'mi439', 'Mi439']
SUCCESS  | utils:image_works_with_device:54 - Device supported by the selected image.
SUCCESS  | utils:recovery_works_with_device:87 - Device supported by the selected recovery.
ERROR    | views.select_view:enable_button_if_ready:625 - Some additional images don't match. Please select different ones.
```
So the problem doesn't seem to be on `pick_vbmeta_result` function, don't know where to search :thinking:

tsterbak commented:
> Hello! I started to test today. I didn't succeed flashing the phone for now.

Hey, thanks for testing!

I updated the PR on your suggestions 1, 3 and 4 as well as hopefully fixed the bug on the vbmeta selection.

I added the other suggestions as todos, since they are straight forward to add, when the process works :)

anon1892 commented:
Okay, nice!

You missed some changes : 
```
Traceback (most recent call last):
  File "/home/******/lib/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/home/******/lib/python3.9/threading.py", line 917, in run
    self._target(*self._args, **self._kwargs)
  File "/home/******/lib/python3.9/site-packages/flet_core/event_handler.py", line 28, in __sync_handler
    h(r)
  File "/home/******/openandroidinstaller/views/select_view.py", line 598, in enable_button_if_ready
    and recovery_works_with_device(
TypeError: recovery_works_with_device() got an unexpected keyword argument 'device_code'
```
There are two occurrences in the file. I just have to replace with `supported_device_codes=self.state.config.supported_device_codes` for the code to works.

Little bug: the `vbmeta.img` checkbox is red when it is ok (but checked, so it works)

Else, I succeeded flashing Lineage 20 on my Mi439.
Bypass of bootloader unlock works. Bypass of recovery flashing works. Maybe we should add a warning, because when the second option is checked, additional partitions are not flashed (so no need to ask for them too)
Also, when recovery flashing is disabled, `adb reboot recovery` is not performed (I think it should be performed all the time, because it works if the phone is already on recovery too)

tsterbak commented:
> There are two occurrences in the file. I just have to replace with supported_device_codes=self.state.config.supported_device_codes for the code to works.

Ah stupid, I forgot that... :grimacing: If fixed it now (hopefully)

> Little bug: the vbmeta.img checkbox is red when it is ok (but checked, so it works)

Should also be fixed now.

> Else, I succeeded flashing Lineage 20 on my Mi439.

Great! :partying_face: :partying_face: 
Then I start implementing the required missing pieces for this :) And I'll try to get a device which needs this feature, then I can also test and you don't have to suffer. :sweat_smile:

tsterbak commented:
@anon1892 Hi, I just update the PR to add the missing pieces I guess. Hopefully everything works as excpected. I tested the "usual" workflow and it still works. Let me know if it works for you.

@MagicLike Feel free to also test it with your devices the find any regressions :)

anon1892 commented on openandroidinstaller/views/select_view.py:
```diff
@@ -296,11 +301,11 @@ def get_notes(self) -> str:
                 "- If something goes wrong, you can reinstall MiUI here:\n<https://xiaomifirmwareupdater.com/>\n"
             )
-        elif brand in "poco":
+        elif brand == "poco":
```
Why two conditions? Poco is Xiaomi, `if brand == "xiaomi" or brand == "poco"` works, isn't it?
And please remove `miui/{self.state.config.device_code}/` in the link, you forget this one ^^

tsterbak replied:
That looks like an issue introduced by merging with main... I corrected it now

anon1892 commented:
Nice, I will test that in a few days!

tsterbak commented:
> Nice, I will test that in a few days!

Great! Thanks a lot! Hopefully I didn't introduce any other issues :D

> Little question, will my device PR (https://github.com/openandroidinstaller-dev/openandroidinstaller/pull/222 and so) will be merged with this update? Without OFox support, isn't it?

If this is possible, it would be great to have them in with TWRP (if TWRP is supported and makes sense). Otherwise I would aim to add Ofox support with the following release and include the devices there. What do you think?

anon1892 commented:
Yes, they work also on TWRP (but some don't have official support for A13 for instance)
OrangeFox was initially created for Xiaomi devices, but TWRP works too.

We just have to remove `supported_recovery` metadata, since it will be added with OFox support, else it already works!

anon1892 commented:
Hello, I tested your changes and found an issue : 
When recovery is already flashed (and the bootloader is unlocked too), there is no button "Confirm and run" to run "adb reboot recovery" : 
![image](https://github.com/openandroidinstaller-dev/openandroidinstaller/assets/40572253/ee6637fd-b858-4774-8f66-041821b7f0df)

And a little improvement: during sideload, the text on top of the progress bar is the same as during selecting addons. Maybe change it for something like "Please be patient, it may take a few minutes. If the process doesn't go to 100% but without errors, no worries; it's normal."

There is no button to allow flashing additional partitions below A13, neither to not flash additional partitions on A13. Is it normal? If in a specific case it is (not) needed, I think the user should be able to choose... 

Else, it seems to be all good!

MagicLike commented:
When the bugs @anon1892 found are fixed, I'll test it with my OnePlus One to see if there are any issues with older devices :+1:

tsterbak commented:
> And a little improvement: during sideload, the text on top of the progress bar is the same as during selecting addons. Maybe change it for something like "Please be patient, it may take a few minutes. If the process doesn't go to 100% but without errors, no worries; it's normal."

I updated the accompanying text to the progressbar now. But it should go to 100% in the end ;) So this part is not there.

>There is no button to allow flashing additional partitions below A13, neither to not flash additional partitions on A13. Is it normal? If in a specific case it is (not) needed, I think the user should be able to choose...

I added a button to skip the additional partitions if you want to. But I think there should be no option to flash additional partitions even if the config doesn't tell you. Maybe there should be an option in the config to specify for which Android versions it should be possible/necessary to flash additional partitions. I want to keep the options as low as reasonably possible on the user side. I understand, that there might be reasons to skip the additional partitions (for example if you already did it), but you should not do it if it is not specified in the config. What do you think?

> Else, it seems to be all good!

Great!

tsterbak commented:
> As of 4/9 , this branch cannot be used to flash any device , even if "recovery is already flashed" is ticked the recovery file is not validated and you cannot go to next step , I will have a look to see what is wrong

Hey, thanks a lot for testing and reporting this issue! It should be fixed and improved now! :)

freechelmi commented:
> > As of 4/9 , this branch cannot be used to flash any device , even if "recovery is already flashed" is ticked the recovery file is not validated and you cannot go to next step , I will have a look to see what is wrong
> 
> Hey, thanks a lot for testing and reporting this issue! It should be fixed and improved now! :)

That was fast !!! Yes it works now

anon1892 commented:
> I updated the accompanying text to the progressbar now. But it should go to 100% in the end ;) So this part is not there.

Ok! I was writing this because adb behaves strangly, and on my phone, stop at 47% when sideload is finished...

> I want to keep the options as low as reasonably possible on the user side.

Maybe show this button only in a “developer mode”?

tsterbak commented:
> Ok! I was writing this because adb behaves strangly, and on my phone, stop at 47% when sideload is finished...

Ah yes, that's true. The installer on the PC reaches 100% at the end but on the phone it only gets to 47%...

> Maybe show this button only in a “developer mode”?

Yeah, I need to prepare to move all "advanced features" to a developer or advanced mode. I'll try to simplify stuff again soon :)

MagicLike commented:
> When the bugs @anon1892 found are fixed, I'll test it with my OnePlus One to see if there are any issues with older devices 👍

Ok, this is weird: My OnePlus One won't boot the twrp I got here (which worked all the time) and also won't flash it... I will try to investigate in the next few weeks, but I am also pretty busy atm...

tsterbak commented:
Thanks for testing. Then I will also do some more testing with older devices, maybe I find something :)

tsterbak commented:
> Ok, this is weird: My OnePlus One won't boot the twrp I got here (which worked all the time) and also won't flash it... I will try to investigate in the next few weeks, but I am also pretty busy atm...

So, I got to test with some old Samsung and Sony devices (so far I don't own a OnePlus device). And as far as I can tell everything seems to work as expected.
Could you reproduce your issue? What did the termial/log output say?

MagicLike commented:
> > Ok, this is weird: My OnePlus One won't boot the twrp I got here (which worked all the time) and also won't flash it... I will try to investigate in the next few weeks, but I am also pretty busy atm...
> 
> So, I got to test with some old Samsung and Sony devices (so far I don't own a OnePlus device). And as far as I can tell everything seems to work as expected. Could you reproduce your issue? What did the terminal/log output say?

Yes, I just reproduced the issue. The OAI log does not show any interesting or relevant information regarding the error - just the normal log...

```
2023-09-15 18:51:08.329 | INFO     | tooling:logging:88 - Boot custom recovery with fastboot. - Parameters: {'recovery': '/*/twrp-3.6.2_9-0-bacon.img', 'is_ab': False, 'bin_path': PosixPath('/*/openandroidinstaller/openandroidinstaller/bin')}
2023-09-15 18:51:08.329 | INFO     | tooling:fastboot_boot_recovery:414 - Boot custom recovery with fastboot.
2023-09-15 18:51:08.329 | INFO     | tooling:run_command:60 - Run command: ['/*/openandroidinstaller/openandroidinstaller/bin/fastboot', 'boot', '/*/twrp-3.6.2_9-0-bacon.img']
2023-09-15 18:51:09.099 | INFO     | tooling:run_command:73 - Sending 'boot.img' (15256 KB)                      OKAY [  0.480s]
2023-09-15 18:51:09.126 | INFO     | tooling:run_command:73 - Booting                                            OKAY [  0.026s]
2023-09-15 18:51:09.127 | INFO     | tooling:run_command:73 - Finished. Total time: 0.535s
2023-09-15 18:51:09.128 | INFO     | tooling:logging:88 - Wait for recovery - Parameters: {'bin_path': PosixPath('/*/openandroidinstaller/openandroidinstaller/bin')}
2023-09-15 18:51:09.128 | INFO     | tooling:run_command:60 - Run command: ['/*/openandroidinstaller/openandroidinstaller/bin/adb', 'wait-for-recovery']
```

But even after waiting minutes, nothing is happening on the OnePlus One side. - It seems like it freezes every time, as fastboot does not respond to fastboot commands anymore.

freechelmi commented:
I have a one plus one and faced the same behaviour on current stable 0.4.5 

The step where we try to boot the recovery via fastboot just let it stuck like you described

MagicLike commented:
> I have a one plus one and faced the same behaviour on current stable 0.4.5
> 
> The step where we try to boot the recovery via fastboot just let it stuck like you described

I also tried to boot twrp manually over cli, but nothing changes...

tsterbak commented:
> I have a one plus one and faced the same behaviour on current stable 0.4.5

And I assume there was no issue with version 0.4.4, right? So there must be a regression somewhere on the `dev` branch. :thinking:

MagicLike commented:
> > I have a one plus one and faced the same behaviour on current stable 0.4.5
> 
> And I assume there was no issue with version 0.4.4, right? So there must be a regression somewhere on the `dev` branch. 🤔

Nope same issue on 0.4.4 - I don't think that this is an issue with OAI, but rather with the devices itself...

tsterbak commented:
> Nope same issue on 0.4.4 - I don't think that this is an issue with OAI, but rather with the devices itself...

Hm, might be. Maybe it does not implement "wait for recovery" yet? :thinking:  What happens if you boot to recovery manually and then run `adb wait-for-recovery` manually?

MagicLike commented:
> > Nope same issue on 0.4.4 - I don't think that this is an issue with OAI, but rather with the devices itself...
> 
> Hm, might be. Maybe it does not implement "wait for recovery" yet? 🤔 What happens if you boot to recovery manually and then run `adb wait-for-recovery` manually?

I can't even boot the recovery - that is the point: It worked perfectly fine weeks ago and suddenly I am only able to flash stock stuff and not even boot custom recoveries. The bootloader just freezes and becomes unavailable.

MagicLike commented:
I have the feeling, that there is some kind of limit reached / the recovery storage sector might be broken or something else...

tsterbak commented:
> I can't even boot the recovery - that is the point: It worked perfectly fine weeks ago and suddenly I am only able to flash stock stuff and not even boot custom recoveries. The bootloader just freezes and becomes unavailable.

Hm, that's weird. I had a bit of a similar thing (for a samsung device) and just using a different usb-cable somehow seemed to fix it :man_shrugging: :thinking:  but could also have been random...

> I have the feeling, that there is some kind of limit reached / the recovery storage sector might be broken or something else...
Maybe it is also soft-bricked...

Never heard of something like that (which does not necessarily means something). Booting the recovery leaves no real trace on the device. And if you can flash the stock recovery and image, then something seems to work still. Really interesting. :thinking:

MagicLike commented:
> Hm, that's weird. I had a bit of a similar thing (for a samsung device) and just using a different usb-cable somehow seemed to fix it 🤷‍♂️ 🤔 but could also have been random...

Tried another cable - no difference :disappointed:

tsterbak commented:
Can we move forward with this? I'm slowly wrapping up stuff for the new release.

What can/should we do about the OnePlus issues? Somehow weird, that it comes up now. Can someone try with an older version of the installer?
